### ISSUE TITLE
Why setting CPU resource requirements to 0 in cache preparation?

### ISSUE THREAD

**Issue by tjingrant** [stanford-crfm/levanter#383]
Hi, thanks for the great work!

I'm curious why you choose to set cpu requirement to 0.0 in a bunch of ray tasks/actors during dataset caching, for example:
https://github.com/stanford-crfm/levanter/blob/133dbba64ff08873ed07a5661211db867a471343/src/levanter/data/shard_cache.py#L254

I've been trying to cache a fairly large dataset (C4) on a reasonably large number of CPUs (slightly more then 1K); every time I run the dataset caching script I run into resource exhaustion issues. Setting the cpu requirements to 1 or 2 generally improves things significantly though it can still lead to resource exhaustion due to ray releasing resource associated with tasks waiting on ray.get().

So I'm curious to know what's the rationales behind resource requirement setting and whether this is potentially a bug.

**Comment by dlwh** [comment]
Hi @tjingrant,

Hrm, that's interesting. I'll be honest I didn't really tune these beyond making sure that a TPU v3-32 (8 machines with ~48 cores) could handle The Pile. The main use case is for preprocessing running in the background while training is running on the CPUs that control training, so I didn't really tune for the 1000s of CPUS in the "offline" case.

I guess I thought it was doing basically no work (besides json parsing), so it was fine to set it to 0. (The tasks that actually do preprocessing have their own resource requirements. (This is what _mk_queue_aware_process_task is doing)

I do need to improve this though. I will try to look into this this week.

**Comment by dlwh** [comment]
i'm working on fixing this right now, but given my history with Ray this is likely to take a few days. Could you try reducing the number of shards to ~32 or ~128 (by e.g. concatenating)? That might unblock you.

**Comment by tjingrant** [comment]
thanks for the quick response, I'll give your suggestion a try later today.

Meanwhile, let me show you the config file I used:
```yaml
train_urls:
  - "https://huggingface.co/datasets/allenai/c4/resolve/1ddc917116b730e1859edef32896ec5c16be51d0/en/c4-train.{00000..01023}-of-01024.json.gz"
cache_dir: "gs://sparse-llm/data_cache/tokenized/c4_llama/"
tokenizer: "meta-llama/Llama-2-7b-hf"
```

I think 10 hours later this is what I see on Ray dashboard; it seems like most tasks are not making progress; this is on 240 cpus (TPU v4-8), I can try TPU v4-32 later too, which will look more like your original setup.

<img width="1619" alt="Screenshot 2023-12-12 at 10 45 44 AM" src="https://github.com/stanford-crfm/levanter/assets/6410074/2435ec16-9e53-4092-a45a-c43bbf748d4d">

**Comment by dlwh** [comment]
I think there's just something wrong if the number of shards is too high
(pile is 32 shards or so). i'm trying to fix

**Comment by tjingrant** [comment]
It appears that at least part of the issue is that the c4 dataset is marked as a json file, but is actually a jsonl file.

Renaming them appears to help relieve some of the memory issue.

But anyhow, for what it's worth, two tricks required to get c4 working are:
- Concat dataset partitions to reduce shards.
- Rename json files in c4 to jsonl.

**Comment by dlwh** [comment]
Oh interesting. That's good to know! I guess I knew that sometimes people write json for jsonl (HF does this a lot). We should try to sniff out json vs jsonl automatically

**Comment by tjingrant** [comment]
> We should try to sniff out json vs jsonl automatically

no promises but I can try to submit a pr for that.

**Comment by dlwh** [comment]
oh thanks!

**Comment by dlwh** [comment]
If you're up for it, could you try this branch`actor_pqueue_stable`? On OWT it's maintaining about 50% cpu utilization, which is fast enough that it will stay ahead of the actual training.

Working on improving it more

**Comment by dlwh** [comment]
I believe this should be fixed in main now that #383 is merged

**Comment by tjingrant** [comment]
@dlwh I believe it's working after fixing the file format sniffing issue with C4:

https://github.com/stanford-crfm/levanter/pull/389
